src/v/cloud_storage/segment_meta_cstore.h

```diff
@@ -0,0 +1,848 @@
+    // Transactional append operation
+    //
+    // The method returns an optional with the state of the
+    // transaction. If the append operation doesn't require
+    // memory allocation the optional is null. Otherwise it
+    // will contain a transaction which need to be commited.
+    // The actual append operation can allocate memory and
+    // can throw but 'commit' method of the transaction
+    // doesn't allocate and doesn't throw.
+    //
+    // Transaction can be aborted by destroying the tx object.
+    // Only one transaction can be active at a time. The caller
+    // of the method is supposed to check the returned value
+    // and to call 'commit' if it's not 'nullopt'.
+    std::optional<column_tx_t> append_tx(value_t value) {
+        std::optional<column_tx_t> tx;
+        if (_frames.empty() || _frames.back().size() == max_frame_size) {
+            frame_t tmp(_delta_alg);
+            tmp.append(value);
+            tx.emplace(std::move(tmp), *this);
+            return tx;
+        }
+        auto inner = _frames.back().append_tx(value);
+        if (!inner.has_value()) {
+            // transactional op is already committed
+            return tx;
+        }
+        tx.emplace(std::move(*inner), *this);
+        return tx;
+    }
+
+    const_iterator at(size_t index) const {
+        auto inner = index;
+        for (auto it = _frames.begin(); it != _frames.end(); ++it) {
+            if (it->size() <= inner) {
+                inner -= it->size();
+            } else {
+                return const_iterator(it, _frames.end(), inner, index);
+            }
+        }
+        return end();
+    }
+
+    /// Get element by index, use 'hint' to speedup the operation.
+    ///
+    /// The 'hint' has to correspond to any index lower or equal to
+    /// 'index'. The 'hint' values has to be stored externally.
+    const_iterator at(size_t index, const hint_t& hint) const {
+        auto inner = index;
+        for (auto it = _frames.begin(); it != _frames.end(); ++it) {
+            if (it->size() <= inner) {
+                inner -= it->size();
+            } else {
+                return const_iterator(it, _frames.end(), inner, hint, index);
+            }
+        }
+        return end();
+    }
+
+    size_t size() const {
+        if (_frames.empty()) {
+            return 0;
+        }
+        size_t total = 0;
+        for (const auto& f : _frames) {
+            total += f.size();
+        }
+        return total;
+    }
+
+    size_t mem_use() const {
+        size_t total = 0;
+        for (const auto& p : _frames) {
+            total += p.mem_use();
+        }
+        return sizeof(*this) + total;
+    }
+
+    const_iterator begin() const { return const_iterator(_frames); }
+
+    const_iterator end() const { return const_iterator(); }
+
+    const_iterator find(value_t value) const {
+        return static_cast<const Derived*>(this)
+          ->template pred_search<std::equal_to<value_t>>(value);
```

Comment:
Do I understand correctly that elements are ordered by offset form large to small, so the pred_search in these cases is actually searching from the front of the log backward?

Whichever direction we're searching in, it feels like this would benefit from some logic to notice when we're e.g. fetching something very close to one end, and start search from that end, given that this is an list of unbounded length (since it's a doubly linked list, it's okay if queries to the middle are slow, but queries to the beginning should be fast + so should queries to near the end)

Reply:
The search is done from smaller offsets to larger offsets.  It's a good idea to leverage search from both ends. I can improve this in the next PR. Maybe even switch from list to map. But it's probably fast enough even in current shape.